components/director/internal/domain/formation/service.go

```diff
@@ -889,64 +881,140 @@ func (s *service) UnassignFormation(ctx context.Context, tnt, objectID string, o
+	err = nerr
 	if err != nil {
 		if apperrors.IsNotFoundError(err) {
-			commitErr := tx.Commit()
-			if commitErr != nil {
-				return nil, errors.Wrapf(err, "while committing transaction with error")
+			if commitErr := tx.Commit(); commitErr != nil {
+				err = errors.Wrapf(
+					errors.Wrapf(commitErr, "while committing transaction"),
+					errors.Wrapf(nerr, "while generating formation assignment notifications").Error(),
+				)
+				return nil, err
 			}
+
 			return formationFromDB, nil
 		}
 		return nil, errors.Wrapf(err, "while generating notifications for %s unassignment", objectType)
 	}
 
-	formationAssignmentsForObject, err := s.formationAssignmentService.ListFormationAssignmentsForObjectID(transactionCtx, formationFromDB.ID, objectID)
-	if err != nil {
-		return nil, errors.Wrapf(err, "while listing formationAssignments for object with type %q and ID %q", objectType, objectID)
+	formationAssignmentClonesForObject := make([]*model.FormationAssignment, 0)
+	for _, ia := range initialAssignmentsData {
+		formationAssignmentClonesForObject = append(formationAssignmentClonesForObject, ia.Clone())
 	}
 
-	rtmContextIDsMapping, err := s.getRuntimeContextIDToRuntimeIDMapping(transactionCtx, tnt, formationAssignmentsForObject)
+	rtmContextIDsMapping, nerr := s.getRuntimeContextIDToRuntimeIDMapping(transactionCtx, tnt, formationAssignmentClonesForObject)
+	err = nerr
 	if err != nil {
 		return nil, err
 	}
 
-	applicationIDToApplicationTemplateIDMapping, err := s.getApplicationIDToApplicationTemplateIDMapping(transactionCtx, tnt, formationAssignmentsForObject)
+	applicationIDToApplicationTemplateIDMapping, nerr := s.getApplicationIDToApplicationTemplateIDMapping(transactionCtx, tnt, formationAssignmentClonesForObject)
+	err = nerr
 	if err != nil {
 		return nil, err
 	}
 
-	if err = s.formationAssignmentService.ProcessFormationAssignments(transactionCtx, formationAssignmentsForObject, rtmContextIDsMapping, applicationIDToApplicationTemplateIDMapping, requests, s.formationAssignmentService.CleanupFormationAssignment, model.UnassignFormation); err != nil {
-		commitErr := tx.Commit()
-		if commitErr != nil {
-			return nil, errors.Wrapf(err, "while committing transaction with error")
+	if nerr = s.formationAssignmentService.ProcessFormationAssignments(transactionCtx, formationAssignmentClonesForObject, rtmContextIDsMapping, applicationIDToApplicationTemplateIDMapping, requests, s.formationAssignmentService.CleanupFormationAssignment, model.UnassignFormation); nerr != nil {
+		err = nerr
+		if commitErr := tx.Commit(); commitErr != nil {
+			err = errors.Wrapf(
+				errors.Wrapf(commitErr, "while committing transaction"),
+				errors.Wrapf(nerr, "while processing formation assignments").Error(),
+			)
+			return nil, err
 		}
+
+		shouldRevertAssignmentsUpdateFromTransaction = false
```

Comment:
Should be after the transaction is commited. Currently if the commit fails the assignments will not be reverted

Reply:
we have another flag that is flipped after the transaction is committed, on line 1004. This flag here is needed to cover the case when we have "processFormationAssignments" error and the transaction is successfully committed. because we'll return with the err on line 994 and we don't want to revert the changed FAs, since the inner transaction is successfully committed.